tests/conftest.py

```diff
@@ -216,18 +237,21 @@ def create_one_fake_project(
             Dict[str, Union[Deps, Tuple[Deps, ExtraDeps]]]
         ] = None,
         extra_file_contents: Optional[Dict[str, str]] = None,
+        is_dynamic: bool = False,
```

Comment:
Thinking some more about how to handle dynamic `pyproject.toml`s: Instead of having an `is_dynamic` flag here, you could instead have a `create_dynamic_pyproject_toml` bool flag. Instead of passing this flag on to `format_deps()` and the underlying formatter function, you can add something like this around line 255:

```python
        if files_with_declared_deps is not None:
            for filename, all_deps in files_with_declared_deps.items():
                tmp_files[filename] = format_deps(filename, all_deps, is_dynamic)
            if create_dynamic_pyproject_toml:
                # Split files_with_declared_deps arbitrarily into regular and optional deps
                # and construct a pyproject.toml with dynamic sections pointing to them.
                req_files = list(files_with_declared_deps.keys())
                deps, optional_deps = req_files[:len(req_files) / 2], req_files[len(req_files) / 2:]
                tmp_files["pyproject.toml"] = f"""\
                    [project]
                    name = "MyLib"
                    dynamic = ["dependencies", "optional-dependencies"]
                    [tool.setuptools.dynamic]
                    dependencies = {{ file = {deps!r} }}
                    optional-dependencies.test = {{ file = {optional_deps!r} }}
                    """
        if extra_file_contents is not None:
            ...
```

Reply:
Having a `create_dynamic_pyproject_toml` flag is a good idea. The only reason why I decided to hardcode the dynamic section under each test at last is that we also need to test the case where the toml file has both `[project.dependencies]` and `[tool.setuptools.dynamic.dependencies]`. If we turn on this `create_dynamic_pyproject_toml` flag, it will be complicated to add some non-dynamic dependencies because the pyproject.toml in the fake project will be replaced. 
Please let me know if hardcoding the dynamic section looks good to you!

Reply:
Yes, hardcoding the `pyproject.toml` in these tests is good with me. Adding a helper to generate it is only useful if it's easier to read the tests _with_ the helper than _without_, and for the number of tests (few) and complexity (high) here, it's not worth it.